code/photometry.py

```diff
@@ -0,0 +1,168 @@
+try:
+    from tractor import (Tractor, PixelizedPSF, NullWCS,
+                         NullPhotoCal, ConstantSky, Image)
+
+except ImportError:
+    print("tractor is missing")
+    pass
+
+
+def calc_background(*, bkgsubimage):
+    """Measure sky noise and mean level.
+
+    Parameters:
+    -----------
+    bkgsubimage: <need type>
+        Image data from which the sky background will be calculated.
+
+    Returns:
+    --------
+    skymean: float
+        Mean of sigma-clipped background
+    skynoise: float
+        Standard deviation of sigma-clipped background
+    """
+    # suppress warnings about nans in the calculation
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=RuntimeWarning)
+        skymean, skymedian, skynoise = sigma_clipped_stats(
+            bkgsubimage, sigma=3.0)
+
+    return skymean, skynoise
+
+
+def run_tractor(*, subimage, prf, objsrc, skymean, skynoise):
+    """Make the tractor image and perform forced photometry.
+
+    Parameters:
+    -----------
+    subimage: <need type>
+        Science image cutout.
+    prf: np.ndarray
+        Point spread function for the band/channel.
+    objsrc: List[tractor.ducks.Source]
+        List of tractor Source objects for the target and nearby sources.
+    skymean: float
+        Mean of sigma-clipped background
+    skynoise: float
+        Standard deviation of sigma-clipped background
+
+    Returns:
+    -------
+    flux_var: float, double, or None
+        Flux variance result from the tractor optimization.
+        None if tractor optimization succeeded but it didn't report a variance.
+
+    Raises:
+    -------
+    TractorError: If the tractor optimization fails.
+    """
+    # make the tractor image
+    tim = Image(
+        data=subimage,
+        invvar=np.ones_like(subimage) / skynoise**2,
+        psf=PixelizedPSF(prf),
+        wcs=NullWCS(),
+        photocal=NullPhotoCal(),
+        sky=ConstantSky(skymean),
+    )
+
+    # make tractor object combining tractor image and source list
+    tractor = Tractor([tim], objsrc)  # [src]
+
+    # freeze the parameters we don't want tractor fitting
+    tractor.freezeParam("images")  # now fits 2 positions and flux
+    # tractor.freezeAllRecursive()#only fit for flux
+    # tractor.thawPathsTo('brightness')
+
+    # run the tractor optimization (do forced photometry)
+    # Take several linearized least squares steps
+    try:
+        tr = 0
+        with suppress_stdout():
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", ".*divide by zero.*")
+                # warnings.simplefilter('ignore')
+                for tr in range(20):
+                    dlnp, X, alpha, flux_var = tractor.optimize(variance=True)
+                    # print('dlnp',dlnp)
+                    if dlnp < 1e-3:
+                        break
+
+    # catch exceptions and bad fits
+    except Exception as e:
```

Comment:
I know that we can't necessarily change this now, but rule of thumb: only except specific exceptions rather than the very generic one. This current approach will catch all bugs and expected issues under one umbrella and calls it convergence error, while it may not be always the case.

Reply:
I understand and agree. To be clear, these try-except blocks are not introduced by this PR, they were already here. The only change I made to them directly was to replace the bare `except`s with `except Exception`. Otherwise, I tried to patch them up a bit around the edges and isolate them from the rest of the code, but I intentionally did not change how they fundamentally work because that would have interfered with my ability to test the changes that were the real focus of this PR.